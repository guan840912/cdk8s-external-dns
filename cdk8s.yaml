language: typescript
app: node ./lib/integ.default.js
imports:
  - k8s