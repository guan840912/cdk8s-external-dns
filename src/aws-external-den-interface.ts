export enum AwsZoneTypeOptions {
  /**
   * will make ExternalDNS see only the hosted zones matching provided domain, omit to process all available hosted zones
   */
  PUBLIC = 'public',
  /**
   * will make ExternalDNS see only the hosted zones matching provided domain, omit to process all available hosted zones
   */
  PRIVATE = 'private',
}