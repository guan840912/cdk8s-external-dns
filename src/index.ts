export * from './aws-external-dns';
export * from './aws-external-dns-policy';
export * from './aws-external-den-interface';