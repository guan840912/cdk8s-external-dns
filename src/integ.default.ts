
import { App, Chart } from 'cdk8s';
import { Construct } from 'constructs';
import { AwsExternalDns, AwsZoneTypeOptions } from './index';

export class MyChart extends Chart {
  constructor(scope: Construct, name: string) {
    super(scope, name);
    new AwsExternalDns(this, 'cdk8sAwsExternalDns', {
      domainFilter: 'exmaple.com',
      awsZoneType: AwsZoneTypeOptions.PUBLIC,
      namespace: 'default',
      serviceAccountName: 'external-dns',
    });
  }
}
const app = new App();
new MyChart(app, 'test');
app.synth();
