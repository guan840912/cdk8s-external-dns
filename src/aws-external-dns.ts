import * as cdk8s from 'cdk8s';
import { Construct } from 'constructs';
import { AwsZoneTypeOptions } from './aws-external-den-interface';

export interface AwsExternalDnsOptions {

  /**
   * Service Account Name for external-dns.
   *
   * @default - external-dns
   */
  readonly serviceAccountName?: string;

  /**
   * Namespace for external-dns
   *
   * @default - default
   */

  readonly namespace?: string;
  /**
   * will make ExternalDNS see only the hosted zones matching provided domain, omit to process all available hosted zones
   *
   * @default - none
   *
   * @example - mydomain.com
   */
  readonly domainFilter: string;

  /**
   * will make ExternalDNS see only the hosted zones matching provided domain, omit to process all available hosted zones
   *
   * @default - AwsZoneTypeOptions.PUBLIC = public
   */
  readonly awsZoneType?: AwsZoneTypeOptions;

  /**
   * image for external-dns.
   * @default - k8s.gcr.io/external-dns/external-dns:v0.7.3
   */
  readonly image?: string;
}

/**
 * Generate external-dns config yaml.
 * see https://github.com/kubernetes-sigs/external-dns/blob/master/docs/tutorials/aws.md
*/
export class AwsExternalDns extends Construct {
  /**
   * Service Account Name for external-dns.
   */
  public readonly serviceAccountName: string;
  /**
   * Kubernetes Deployment Name for external-dns.
   */
  public readonly deploymentName: string;
  /**
   * Namespace for external-dns.
   * @default - default
   */
  public readonly namespace: string ;
  /**
   * image for external-dns.
   * @default - k8s.gcr.io/external-dns/external-dns:v0.7.3
   */
  public readonly image: string ;
  /**
   * will make ExternalDNS see only the hosted zones matching provided domain, omit to process all available hosted zones
   *
   * @default - AwsZoneTypeOptions.PUBLIC = public
   */
  public readonly awsZoneType: AwsZoneTypeOptions;

  constructor(scope: Construct, id: string, options: AwsExternalDnsOptions) {
    super(scope, id);
    this.serviceAccountName = options.serviceAccountName ?? 'external-dns';
    this.deploymentName = 'external-dns';
    this.namespace = options.namespace ?? 'default';
    this.image = options.image ?? 'k8s.gcr.io/external-dns/external-dns:v0.7.3';
    this.awsZoneType = options.awsZoneType ?? AwsZoneTypeOptions.PUBLIC;
    new cdk8s.ApiObject(this, 'external-dns-cluster-role', {
      apiVersion: 'rbac.authorization.k8s.io/v1beta1',
      kind: 'ClusterRole',
      metadata: {
        name: 'external-dns',
      },
      rules: [
        {
          apiGroups: [
            '',
          ],
          resources: [
            'services',
            'endpoints',
            'pods',
          ],
          verbs: [
            'get',
            'watch',
            'list',
          ],
        },
        {
          apiGroups: [
            'extensions',
            'networking.k8s.io',
          ],
          resources: [
            'ingresses',
          ],
          verbs: [
            'get',
            'watch',
            'list',
          ],
        },
        {
          apiGroups: [
            '',
          ],
          resources: [
            'nodes',
          ],
          verbs: [
            'list',
            'watch',
          ],
        },
      ],
    });
    new cdk8s.ApiObject(this, 'external-dns-cluster-role-binding', {
      apiVersion: 'rbac.authorization.k8s.io/v1beta1',
      kind: 'ClusterRoleBinding',
      metadata: {
        name: 'external-dns-viewer',
      },
      roleRef: {
        apiGroup: 'rbac.authorization.k8s.io',
        kind: 'ClusterRole',
        name: 'external-dns',
      },
      subjects: [
        {
          kind: 'ServiceAccount',
          name: this.serviceAccountName,
          namespace: this.namespace,
        },
      ],
    });

    new cdk8s.ApiObject(this, 'external-dns-deploy', {
      apiVersion: 'apps/v1',
      kind: 'Deployment',
      metadata: {
        name: this.deploymentName,
        namespace: this.namespace,
      },
      spec: {
        strategy: {
          type: 'Recreate',
        },
        selector: {
          matchLabels: {
            app: this.deploymentName,
          },
        },
        template: {
          metadata: {
            labels: {
              app: this.deploymentName,
            },
          },
          spec: {
            serviceAccountName: this.serviceAccountName,
            containers: [
              {
                name: this.deploymentName,
                image: this.image,
                args: [
                  '--source=service',
                  '--source=ingress',
                  '--domain-filter=' + options.domainFilter,
                  '--provider=aws',
                  '--policy=upsert-only',
                  '--aws-zone-type=' + this.awsZoneType,
                  '--registry=txt',
                  '--txt-owner-id=external-dns',
                ],
              },
            ],
            securityContext: {
              fsGroup: 65534,
            },
          },
        },
      },
    });
  }
}
