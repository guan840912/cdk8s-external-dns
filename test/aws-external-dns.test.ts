import { Role, ServicePrincipal } from '@aws-cdk/aws-iam';
import * as cdk from '@aws-cdk/core';
import '@aws-cdk/assert/jest';
import { Chart, Testing } from 'cdk8s';
import { AwsExternalDns, AwsExternalDnsPolicyHelper, AwsZoneTypeOptions } from '../src/index';

test('AwsExternalDns', () => {
  const app = Testing.app();
  const chart = new Chart(app, 'test');
  new AwsExternalDns(chart, 'cdk8sAwsExternalDns', {
    domainFilter: 'exmaple.com',
    awsZoneType: AwsZoneTypeOptions.PRIVATE,
    namespace: 'kube-system',
    serviceAccountName: 'external-dns',
    image: 'k8s.gcr.io/external-dns/external-dns:v0.7.3',
  });
  expect(Testing.synth(chart)).toMatchSnapshot();
});

test('AwsExternalDnsPolicy helper', () => {
  const app = new cdk.App();
  const stack = new cdk.Stack(app, 'test');
  const sa = new Role(stack, 'sarole', {
    assumedBy: new ServicePrincipal('sns.amazonaws.com'),
  });
  AwsExternalDnsPolicyHelper.addPolicy(sa);
  expect(stack).toHaveResource('AWS::IAM::Policy', {
    PolicyDocument: {
      Statement: [
        {
          Action: 'route53:ChangeResourceRecordSets',
          Effect: 'Allow',
          Resource: 'arn:aws:route53:::hostedzone/*',
        },
        {
          Action: [
            'route53:ListHostedZones',
            'route53:ListResourceRecordSets',
          ],
          Effect: 'Allow',
          Resource: '*',
        },
      ],
      Version: '2012-10-17',
    },
  });
});