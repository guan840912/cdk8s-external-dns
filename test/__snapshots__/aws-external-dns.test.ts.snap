// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AwsExternalDns 1`] = `
Array [
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1beta1",
    "kind": "ClusterRole",
    "metadata": Object {
      "name": "external-dns",
    },
    "rules": Array [
      Object {
        "apiGroups": Array [
          "",
        ],
        "resources": Array [
          "services",
          "endpoints",
          "pods",
        ],
        "verbs": Array [
          "get",
          "watch",
          "list",
        ],
      },
      Object {
        "apiGroups": Array [
          "extensions",
          "networking.k8s.io",
        ],
        "resources": Array [
          "ingresses",
        ],
        "verbs": Array [
          "get",
          "watch",
          "list",
        ],
      },
      Object {
        "apiGroups": Array [
          "",
        ],
        "resources": Array [
          "nodes",
        ],
        "verbs": Array [
          "list",
          "watch",
        ],
      },
    ],
  },
  Object {
    "apiVersion": "rbac.authorization.k8s.io/v1beta1",
    "kind": "ClusterRoleBinding",
    "metadata": Object {
      "name": "external-dns-viewer",
    },
    "roleRef": Object {
      "apiGroup": "rbac.authorization.k8s.io",
      "kind": "ClusterRole",
      "name": "external-dns",
    },
    "subjects": Array [
      Object {
        "kind": "ServiceAccount",
        "name": "external-dns",
        "namespace": "kube-system",
      },
    ],
  },
  Object {
    "apiVersion": "apps/v1",
    "kind": "Deployment",
    "metadata": Object {
      "name": "external-dns",
      "namespace": "kube-system",
    },
    "spec": Object {
      "selector": Object {
        "matchLabels": Object {
          "app": "external-dns",
        },
      },
      "strategy": Object {
        "type": "Recreate",
      },
      "template": Object {
        "metadata": Object {
          "labels": Object {
            "app": "external-dns",
          },
        },
        "spec": Object {
          "containers": Array [
            Object {
              "args": Array [
                "--source=service",
                "--source=ingress",
                "--domain-filter=exmaple.com",
                "--provider=aws",
                "--policy=upsert-only",
                "--aws-zone-type=private",
                "--registry=txt",
                "--txt-owner-id=external-dns",
              ],
              "image": "k8s.gcr.io/external-dns/external-dns:v0.7.3",
              "name": "external-dns",
            },
          ],
          "securityContext": Object {
            "fsGroup": 65534,
          },
          "serviceAccountName": "external-dns",
        },
      },
    },
  },
]
`;
